File: Cpp/Introductory Problems/Creating_Strings.cpp
#include <bits/stdc++.h>
using namespace std;

string word;

unordered_map<int, int> facto = {
    {0, 1},
    {1, 1},
    {2, 2},
    {3, 6},
    {4, 24},
    {5, 120},
    {6, 720},
    {7, 5040},
    {8, 40320}
};

string nextPermutation(string &word){
    int l=word.size();
    int index = l-1;
    while (index > 0 && word[index] <= word[index-1]) index--;
    if (index>0){
        int k = l-1;
        while (word[k] <= word[index-1]) k--;
        swap(word[k],word[index-1]);
        int left = index;
        int right = l-1;

        while (left < right){
            swap(word[left],word[right]);
            left++;
            right--;
        }
    }
    else {
        for (int i=1;i<l/2;i++){
            swap(word[i],word[l-1-i]);
        }
    }
    return word;
}

int main(){
    cin >> word;
    sort(word.begin(), word.end());
    int n = word.size();
    int ans = facto[n];

    unordered_map<char,int> counter;
    for (char c:word) counter[c]++;
    for (const auto& pair : counter) ans /= facto[pair.second];
    printf("%d\n",ans);

    for (int i=1; i<=ans;i++){
        cout << word << endl;
        word = nextPermutation(word);
    }
}

File: Python/Introductory Problems/Creating_Strings.py
from collections import Counter

string = list(input())
string.sort()
n = len(string)

facto = {0:1, 1:1, 2:2, 3:6, 4:24, 5:120, 6:720, 7:5040, 8:40320}


# Count number of permutations
count = Counter(string)
ans = facto[n]
for key in count:
    ans = ans//facto[count[key]]

############ next permutation !

def next_permutation(nums):
    l,index = len(nums),len(nums)-1
    while index > 0 and nums[index] <= nums[index-1]:
        index -= 1
    
    if index > 0 :
        # look for smallest number in range index,l st the nb is > cur index
        k = l-1 
        while nums[k] <= nums[index-1]:
            k -= 1
        # swap
        nums[k], nums[index-1] = nums[index-1], nums[k]

        # sort the right part [index,n-1]
        left,right = index,l-1
        while left <= right :
            nums[left],nums[right] = nums[right],nums[left]
            left += 1
            right -= 1
    
    else : # the array is decreasing here
        for i in range(l//2):
            nums[i],nums[l-i-1] = nums[l-1-i],nums[i]
    
    return nums


print(ans)

for i in range(ans):
    print("".join(string))
    string = next_permutation(string)